# Embedding Model Evaluation

In this notebook we will create a RAG pipeline on the mortgage documents provided by Outomation. We are using one standard PDF and one scanned PDF as our corpus for RAG. We will then evaluate different open source embedding models and analyze which one is better suited for our RAG pipeline.

In [ ]:
!pip install llama-index-core llama-index-readers-file llama-index-llms-ollama llama-index-embeddings-huggingface
!pip install llama_index.llms.gemini
!pip install transformers sentence-transformers
!pip install pypdf
!pip install nest_asyncio
!pip install llama_index
!pip install llama-index-experimental
!pip install llama-index-retrievers-bm25
!pip install pytesseract pdf2image
!apt-get update
!apt-get install -y poppler-utils
!pip install pdf2image

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.6/7.6 MB 104.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.8/51.8 kB 4.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.5/42.5 kB 4.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 310.5/310.5 kB 29.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 67.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.9/50.9 kB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 144.4/144.4 kB 12.7 MB/s eta 0:00:00
  Attempting uninstall: setuptools
    Found existing installation: setuptools 75.2.0
    Uninstalling setuptools-75.2.0:
      Successfully uninstalled setuptools-75.2.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
ipython 7.34.0 requires jedi>=0.16, which is not installed.


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.5/4.5 MB 94.0 MB/s eta 0:00:00
  Attempting uninstall: pillow
    Found existing installation: pillow 11.3.0
    Uninstalling pillow-11.3.0:
      Successfully uninstalled pillow-11.3.0


INFO: pip is looking at multiple versions of llama-cloud-services to determine which version is compatible with other requirements. This could take a while.
INFO: pip is still looking at multiple versions of llama-cloud-services to determine which version is compatible with other requirements. This could take a while.
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. See https://pip.pypa.io/warnings/backtracking for guidance. If you want to abort this run, press Ctrl + C.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 303.3/303.3 kB 22.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.9/63.9 kB 7.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 kB 8.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.4/44.4 kB 4.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 440.5/440.5 kB 32.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [ ]:
import os
import nest_asyncio
nest_asyncio.apply()

GOOGLE_API_KEY = "AIzaSyBcaa2hsV5pfYUmdi0qL5SgWZIY75VX3ao"  # Replace with your actual API key
os.environ["GOOGLE_API_KEY"] = GOOGLE_API_KEY

In [ ]:
cd /content/drive/MyDrive/Extern_Outomation

/content/drive/MyDrive/Extern_Outomation


In [ ]:
from llama_index.core import SimpleDirectoryReader

documents = SimpleDirectoryReader(input_dir="./basic_rag", required_exts=[".pdf"]).load_data()
print(f"Loaded {len(documents)} document(s).")

Loaded 5 document(s).


In [ ]:
print(documents[0].text[:1000])  # Print the first 1000 characters of the first document

Your actual rate, payment, and cost could be higher. Get an official Loan Estimate before choosing a loan.
Fee Details and Summary
Applicants: Application No:
Date Prepared:
Loan Program:
Prepared By:
THIS IS NOT A GOOD FAITH ESTIMATE (GFE). This "Fees W orksheet" is provided for informational purposes ONLY, to assist
you in determining an estimate of cash that may be required to close and an estimate of your proposed monthly mortgage 
payment. Actual charges may be more or less, and your transaction may not involve a fee for every item listed.
Total Loan Amount:  Interest Rate: Term/Due In:
Fee Paid To Paid By (Fee Split**) Amount PFC / F / POC
TOTAL ESTIMATED FUNDS NEEDED TO CLOSE: TOTAL ESTIMATED MONTHLY PAYMENT:
Total Estimated Funds Total Monthly Payment
Purchase Price (+)
Alterations (+)
Land (+)
Refi (incl. debts to be paid off) (+)
Est. Prepaid Items/Reserves (+)
Est. Closing Costs (+)
Loan Amount (-) Principal & Interest
Other Financing (P & I)
Hazard Insurance
Real Estate Tax

In [ ]:
#we have one scanned pdf in the selected folder so we will convert it to text
from pdf2image import convert_from_path
import pytesseract

pages = convert_from_path('/content/drive/MyDrive/Extern_Outomation/basic_rag/MTG_10009588.pdf')
text = ""
for page in pages:
    text += pytesseract.image_to_string(page)
print(text[:1000])  # Preview extracted text


- QML

MORTGAGE DOC.# 10009588

DOCUMENT NUMBER

RECORDED 06/28/2011 09:35AM
JOHN LA FAVE
NAME & RETURN ADDRESS REGISTER DF DEEDS

M&I Home Lending Solutions
Attn: Secondary Marketing
4121 NW Urbandale Drive
Urbandale, IA 50322

Milwaukee County, WI]
AMOUNT : 30.00
FEE EXEMPT #:

PARCEL IDENTIFIER NUMBER
716-0027-6
[Space Above This Line For Recording Data]

 

State of Wisconsin
581-4247085-703

 

MIN 100273100009309945

THIS MORTGAGE ("Security Instrument") 1s given on June 20, 2011
The Mortgagor is KIMBERLY HOGAN, A Single Person,

("Borrower") This Security Instrument 1s given to Mortgage Electronic Registration Systems, Inc ("MERS"),
(solely as nominee for Lender, as hereinafter defined, and Lender's successors and assigns), as mortgagee MERS 1s
organized and existing under the laws of Delaware, and has an address and telephone number of PO Box .2026,
Flint, MI 48501-2026, tel (888) 679-MERS M&I Bank FSB :
("Lender") 1s organized and existing under the laws of the United States o

In [ ]:
from llama_index.core import Document
doc = Document(text=text, metadata={"file_name": "MTG_10009588.pdf"}) #appending scanned document
documents.append(doc)

In [ ]:
documents_new = []

for doc in documents:
  if doc.text != "":
    documents_new.append(doc)

documents = documents_new

In [ ]:
len(documents)

2

In [ ]:
from llama_index.core.node_parser import SemanticSplitterNodeParser
from llama_index.embeddings.huggingface import HuggingFaceEmbedding

embed_model = HuggingFaceEmbedding(model_name="intfloat/e5-small-v2") #trying different embedding models here
semantic_splitter = SemanticSplitterNodeParser(embed_model=embed_model)

In [ ]:
semantic_nodes = semantic_splitter.get_nodes_from_documents(documents)

In [ ]:
print(f'Total number of semantic nodes: {len(semantic_nodes)}')
print(semantic_nodes[-1])

Total number of semantic nodes: 4
Node ID: 7a18d4f2-7f16-4727-9222-b94cd41959ab
Text: Borrower shall pay when due the principal of, and interest on,
the debt evidenced by the Note and late charges due under the Note  2.
Monthly Payment of Taxes, Insurance and Other Charges. Borrower shall
mclude mm each monthly payment, together with the principal and
interest as set forth in the Note and any late charges, a sum for (a)
taxes and...


We used semantic nodes initially but after improving the rag pipeline in other parts semantic nodes were not providing all the test answers properly because of low number of documents to process leading us to use default chunk splitting method.

In [ ]:
from llama_index.core import VectorStoreIndex
from llama_index.llms.gemini import Gemini

# Create an index with our embeddings
index = VectorStoreIndex.from_documents(documents, embed_model=embed_model)

llm = Gemini(model="models/gemini-2.5-flash")

/tmp/ipython-input-3997311712.py:7: DeprecationWarning: Call to deprecated class Gemini. (Should use `llama-index-llms-google-genai` instead, using Google's latest unified SDK. See: https://docs.llamaindex.ai/en/stable/examples/llm/google_genai/)
  llm = Gemini(model="models/gemini-2.5-flash")


In [ ]:
#checking the index nodes again
print(list(index.docstore.docs.values())[0].text)

Your actual rate, payment, and cost could be higher. Get an official Loan Estimate before choosing a loan.
Fee Details and Summary
Applicants: Application No:
Date Prepared:
Loan Program:
Prepared By:
THIS IS NOT A GOOD FAITH ESTIMATE (GFE). This "Fees W orksheet" is provided for informational purposes ONLY, to assist
you in determining an estimate of cash that may be required to close and an estimate of your proposed monthly mortgage 
payment. Actual charges may be more or less, and your transaction may not involve a fee for every item listed.
Total Loan Amount:  Interest Rate: Term/Due In:
Fee Paid To Paid By (Fee Split**) Amount PFC / F / POC
TOTAL ESTIMATED FUNDS NEEDED TO CLOSE: TOTAL ESTIMATED MONTHLY PAYMENT:
Total Estimated Funds Total Monthly Payment
Purchase Price (+)
Alterations (+)
Land (+)
Refi (incl. debts to be paid off) (+)
Est. Prepaid Items/Reserves (+)
Est. Closing Costs (+)
Loan Amount (-) Principal & Interest
Other Financing (P & I)
Hazard Insurance
Real Estate Tax

In [ ]:
from llama_index.core import Settings
from llama_index.core.query_engine import RetrieverQueryEngine
from llama_index.core.retrievers import QueryFusionRetriever
from llama_index.retrievers.bm25 import BM25Retriever
from llama_index.core.retrievers import BaseRetriever
from llama_index.core.postprocessor import SentenceTransformerRerank


Settings.llm = llm


# Function to create a query engine that uses query expansion plus hybrid and reranking
def build_rag_pipeline(index, llm):

    nodes = list(index.docstore.docs.values())

    # Determine safe top_k value (number of nodes to retrieve)
    # Must be at least 1 and no more than the number of available nodes
    num_nodes = len(nodes)
    safe_top_k = min(3, max(1, num_nodes))

    print(f"Index contains {num_nodes} nodes, using top_k={safe_top_k}")

    vector_retriever = index.as_retriever(
          similarity_top_k = safe_top_k  # Retrieve top 3 most similar chunks
      )

    # Create hybrid retriever (vector + BM25)
    bm25_retriever = BM25Retriever.from_defaults(
        nodes=nodes,
        similarity_top_k=safe_top_k  # Retrieve top 3 most similar chunks
    )

    # Create a proper hybrid retriever class
    class HybridRetriever(BaseRetriever):
        """Hybrid retriever that combines vector and keyword search results."""

        def __init__(self, vector_retriever, keyword_retriever, top_k=3):
            """Initialize with vector and keyword retrievers."""
            self.vector_retriever = vector_retriever
            self.keyword_retriever = keyword_retriever
            self.top_k = top_k
            super().__init__()

        def _retrieve(self, query_bundle, **kwargs):
            """Retrieve from both retrievers and combine results."""
            # Get results from both retrievers
            vector_nodes = self.vector_retriever.retrieve(query_bundle)
            keyword_nodes = self.keyword_retriever.retrieve(query_bundle)

            # Combine all nodes
            all_nodes = list(vector_nodes) + list(keyword_nodes)

            # Remove duplicates (by node_id)
            unique_nodes = {}
            for node in all_nodes:
                if node.node_id not in unique_nodes:
                    unique_nodes[node.node_id] = node

            # Sort by score (higher is better)
            sorted_nodes = sorted(
                unique_nodes.values(),
                key=lambda x: x.score if hasattr(x, 'score') else 0.0,
                reverse=True
            )

            return sorted_nodes[:self.top_k]  # Return top results

    # Create our hybrid retriever instance
    hybrid_retriever = HybridRetriever(
        vector_retriever = vector_retriever,
        keyword_retriever = bm25_retriever,
        top_k=safe_top_k
    )

    # Use QueryFusionRetriever with the hybrid retriever
    fusion_retriever = QueryFusionRetriever(
        retrievers = [hybrid_retriever],
        llm = llm,
        similarity_top_k = 3,
        num_queries = 3,
        mode="reciprocal_rerank"
    )

    # Apply reranking
    reranker = SentenceTransformerRerank(
        model="cross-encoder/ms-marco-MiniLM-L-2-v2",
        top_n=3
    )


    # Plug into query engine
    from llama_index.core.query_engine import RetrieverQueryEngine
    query_engine = RetrieverQueryEngine.from_args(
        retriever = fusion_retriever,
        llm=llm,
        node_postprocessors = [reranker],
        verbose = True
    )
    return query_engine

In [ ]:
rag_engine = build_rag_pipeline(index, llm)

DEBUG:bm25s:Building index from IDs objects


Index contains 4 nodes, using top_k=3


## Checking different Embedding models

MINI LM

In [ ]:
# mini lm
response = rag_engine.query("What is the total estimated monthly payment?")
print('\nFinal Response:\n ---------------------- \n')
print(response)


Final Response:
 ---------------------- 

The total estimated monthly payment is $1,869.37.


In [ ]:
for node in response.source_nodes:
  print(node)

Node ID: c485a76d-1a19-4ac1-9f24-0f398908b631
Text: Your actual rate, payment, and cost could be higher. Get an
official Loan Estimate before choosing a loan. Fee Details and Summary
Applicants: Application No: Date Prepared: Loan Program: Prepared By:
THIS IS NOT A GOOD FAITH ESTIMATE (GFE). This "Fees W orksheet" is
provided for informational purposes ONLY, to assist you in determining
an estim...
Score: -2.472

Node ID: c3290ea6-4005-45be-8b60-88bdb5585b98
Text: Payment of Principal, Interest and Late Charge. Borrower shall
pay when due the principal of, and interest on, the debt evidenced by
the Note and late charges due under the Note  2. Monthly Payment of
Taxes, Insurance and Other Charges. Borrower shall mclude mm each
monthly payment, together with the principal and interest as set forth
in the No...
Score: -3.479

Node ID: 3eaeb641-0a7b-4da5-9bab-edd74f29a25e
Text: - QML  MORTGAGE DOC.# 10009588  DOCUMENT NUMBER  RECORDED
06/28/2011 09:35AM JOHN LA FAVE NAME & RETURN ADDRESS 

In [ ]:
# mini lm
response = rag_engine.query("What is the total estimated monthly payment?")
print('\nFinal Response:\n ---------------------- \n')
print(response)


Final Response:
 ---------------------- 

The borrower pays $650.00 for lender's title insurance.


In [ ]:
for node in response.source_nodes:
  print(node)

Node ID: c3290ea6-4005-45be-8b60-88bdb5585b98
Text: Payment of Principal, Interest and Late Charge. Borrower shall
pay when due the principal of, and interest on, the debt evidenced by
the Note and late charges due under the Note  2. Monthly Payment of
Taxes, Insurance and Other Charges. Borrower shall mclude mm each
monthly payment, together with the principal and interest as set forth
in the No...
Score:  1.785

Node ID: c485a76d-1a19-4ac1-9f24-0f398908b631
Text: Your actual rate, payment, and cost could be higher. Get an
official Loan Estimate before choosing a loan. Fee Details and Summary
Applicants: Application No: Date Prepared: Loan Program: Prepared By:
THIS IS NOT A GOOD FAITH ESTIMATE (GFE). This "Fees W orksheet" is
provided for informational purposes ONLY, to assist you in determining
an estim...
Score:  1.500

Node ID: 8911f1a3-69c5-4564-8454-a19a9945878c
Text: which has the address of 6468 SOUTH 20TH STREET [Street]
MILWAUKEE [City], Wisconsin 53221 [Zip Code] ("Property

In [ ]:
response = rag_engine.query("What are the charges?")
print('\nFinal Response:\n ---------------------- \n')
print(response)


Final Response:
 ---------------------- 

The charges include principal, interest, and late charges on the debt. Monthly payments also encompass sums for taxes and special assessments, leasehold payments or ground rents, and insurance premiums. Additionally, a mortgage insurance premium or a monthly charge in lieu of it may be required.

Other specific charges listed are:
*   Underwriting Fee
*   Wire Transfer Fee
*   Administration Fee
*   Appraisal Fee
*   Credit Report Fee
*   Tax Service Fee
*   Flood Certification Fee
*   Closing/Escrow Fee
*   Document Preparation Fee
*   Notary Fee
*   Lender's Title Insurance
*   Title - Courier Fee
*   Electronic Document Delivery Fee
*   Pest Inspection Fee
*   Home Inspection
*   Mortgage Recording Charge
*   Daily Interest Charges
*   Hazard Insurance Premium


In [ ]:
for node in response.source_nodes:
  print(node)

Node ID: c3290ea6-4005-45be-8b60-88bdb5585b98
Text: Payment of Principal, Interest and Late Charge. Borrower shall
pay when due the principal of, and interest on, the debt evidenced by
the Note and late charges due under the Note  2. Monthly Payment of
Taxes, Insurance and Other Charges. Borrower shall mclude mm each
monthly payment, together with the principal and interest as set forth
in the No...
Score: -5.178

Node ID: c485a76d-1a19-4ac1-9f24-0f398908b631
Text: Your actual rate, payment, and cost could be higher. Get an
official Loan Estimate before choosing a loan. Fee Details and Summary
Applicants: Application No: Date Prepared: Loan Program: Prepared By:
THIS IS NOT A GOOD FAITH ESTIMATE (GFE). This "Fees W orksheet" is
provided for informational purposes ONLY, to assist you in determining
an estim...
Score: -6.780

Node ID: 8911f1a3-69c5-4564-8454-a19a9945878c
Text: which has the address of 6468 SOUTH 20TH STREET [Street]
MILWAUKEE [City], Wisconsin 53221 [Zip Code] ("Property

In [ ]:
query = "What is the maximum loan amount a borrower can apply for?"
response = rag_engine.query(query)

print(response)

The provided context does not specify the maximum loan amount a borrower can apply for. It includes a "Fees Worksheet" with a "Total Loan Amount" of $380,000 for a specific application, but this document is for informational purposes only and does not indicate a maximum limit.


In [ ]:
for node in response.source_nodes:
  print(node)

Node ID: c485a76d-1a19-4ac1-9f24-0f398908b631
Text: Your actual rate, payment, and cost could be higher. Get an
official Loan Estimate before choosing a loan. Fee Details and Summary
Applicants: Application No: Date Prepared: Loan Program: Prepared By:
THIS IS NOT A GOOD FAITH ESTIMATE (GFE). This "Fees W orksheet" is
provided for informational purposes ONLY, to assist you in determining
an estim...
Score: -2.019

Node ID: c3290ea6-4005-45be-8b60-88bdb5585b98
Text: Payment of Principal, Interest and Late Charge. Borrower shall
pay when due the principal of, and interest on, the debt evidenced by
the Note and late charges due under the Note  2. Monthly Payment of
Taxes, Insurance and Other Charges. Borrower shall mclude mm each
monthly payment, together with the principal and interest as set forth
in the No...
Score: -2.245

Node ID: 8911f1a3-69c5-4564-8454-a19a9945878c
Text: which has the address of 6468 SOUTH 20TH STREET [Street]
MILWAUKEE [City], Wisconsin 53221 [Zip Code] ("Property

In [ ]:
response = rag_engine.query("What are the addresses in the document?")
print('\nFinal Response:\n ---------------------- \n')
print(response)


Final Response:
 ---------------------- 

The addresses mentioned are:

*   4121 NW Urbandale Drive, Urbandale, IA 50322
*   PO Box 2026, Flint, MI 48501-2026
*   3993 Howard Hughes Parkway, Las Vegas, NV 89109
*   6468 SOUTH 20TH STREET, MILWAUKEE, Wisconsin 53221


In [ ]:
response = rag_engine.query("Who is the borrower, what is the total loan amount and what is the property Address?")
print('\nFinal Response:\n ---------------------- \n')
print(response)


Final Response:
 ---------------------- 

The borrower is KIMBERLY HOGAN. The total loan amount is $380,000. The property address is 6468 SOUTH 20TH STREET MILWAUKEE, Wisconsin 53221.


BGE Small

In [ ]:
response = rag_engine.query("What is the total estimated monthly payment?")
print('\nFinal Response:\n ---------------------- \n')
print(response)

print('\n')
for node in response.source_nodes:
  print(node)

response = rag_engine.query("How much does the borrower pay for lender's title insurance?")
print('\nFinal Response:\n ---------------------- \n')
print(response)

print('\n')
for node in response.source_nodes:
  print(node)

response = rag_engine.query("What are the charges?")
print('\nFinal Response:\n ---------------------- \n')
print(response)

print('\n')
for node in response.source_nodes:
  print(node)

response = rag_engine.query("Who is the borrower, what is the total loan amount and what is the property Address?")
print('\nFinal Response:\n ---------------------- \n')
print(response)

print('\n')
for node in response.source_nodes:
  print(node)


Final Response:
 ---------------------- 

The total estimated monthly payment is $1,869.37.


Node ID: 9b64fe54-43d5-49e5-b714-cd3910d2850d
Text: Your actual rate, payment, and cost could be higher. Get an
official Loan Estimate before choosing a loan. Fee Details and Summary
Applicants: Application No: Date Prepared: Loan Program: Prepared By:
THIS IS NOT A GOOD FAITH ESTIMATE (GFE). This "Fees W orksheet" is
provided for informational purposes ONLY, to assist you in determining
an estim...
Score: -2.472

Node ID: f794f63a-4a9d-48f0-824a-20d622dbf89f
Text: Payment of Principal, Interest and Late Charge. Borrower shall
pay when due the principal of, and interest on, the debt evidenced by
the Note and late charges due under the Note  2. Monthly Payment of
Taxes, Insurance and Other Charges. Borrower shall mclude mm each
monthly payment, together with the principal and interest as set forth
in the No...
Score: -3.479

Node ID: 66ef84cd-cdb7-4022-b817-d6ab5c0ca46e
Text: - QML  MORTGAGE D

e5 small v2

In [ ]:
response = rag_engine.query("What is the total estimated monthly payment?")
print('\nFinal Response:\n ---------------------- \n')
print(response)

print('\n')
for node in response.source_nodes:
  print(node)

response = rag_engine.query("How much does the borrower pay for lender's title insurance?")
print('\nFinal Response:\n ---------------------- \n')
print(response)

print('\n')
for node in response.source_nodes:
  print(node)

response = rag_engine.query("What are the charges?")
print('\nFinal Response:\n ---------------------- \n')
print(response)

print('\n')
for node in response.source_nodes:
  print(node)

response = rag_engine.query("Who is the borrower, what is the total loan amount and what is the property Address?")
print('\nFinal Response:\n ---------------------- \n')
print(response)

print('\n')
for node in response.source_nodes:
  print(node)


Final Response:
 ---------------------- 

The total estimated monthly payment is $2,308.95.


Node ID: 87bed5a6-5aa8-4410-a28f-82e51d9a9f3d
Text: Your actual rate, payment, and cost could be higher. Get an
official Loan Estimate before choosing a loan. Fee Details and Summary
Applicants: Application No: Date Prepared: Loan Program: Prepared By:
THIS IS NOT A GOOD FAITH ESTIMATE (GFE). This "Fees W orksheet" is
provided for informational purposes ONLY, to assist you in determining
an estim...
Score: -2.472

Node ID: 36099f7d-e9f4-4a06-97ad-36e74de44600
Text: Payment of Principal, Interest and Late Charge. Borrower shall
pay when due the principal of, and interest on, the debt evidenced by
the Note and late charges due under the Note  2. Monthly Payment of
Taxes, Insurance and Other Charges. Borrower shall mclude mm each
monthly payment, together with the principal and interest as set forth
in the No...
Score: -3.479

Node ID: a1131f64-102c-4c4f-94a3-ba6ee8df8d87
Text: - QML  MORTGAGE D

In [ ]:
for node in response.source_nodes:
  print(node.get_text())

- QML

MORTGAGE DOC.# 10009588

DOCUMENT NUMBER

RECORDED 06/28/2011 09:35AM
JOHN LA FAVE
NAME & RETURN ADDRESS REGISTER DF DEEDS

M&I Home Lending Solutions
Attn: Secondary Marketing
4121 NW Urbandale Drive
Urbandale, IA 50322

Milwaukee County, WI]
AMOUNT : 30.00
FEE EXEMPT #:

PARCEL IDENTIFIER NUMBER
716-0027-6
[Space Above This Line For Recording Data]

 

State of Wisconsin
581-4247085-703

 

MIN 100273100009309945

THIS MORTGAGE ("Security Instrument") 1s given on June 20, 2011
The Mortgagor is KIMBERLY HOGAN, A Single Person,

("Borrower") This Security Instrument 1s given to Mortgage Electronic Registration Systems, Inc ("MERS"),
(solely as nominee for Lender, as hereinafter defined, and Lender's successors and assigns), as mortgagee MERS 1s
organized and existing under the laws of Delaware, and has an address and telephone number of PO Box .2026,
Flint, MI 48501-2026, tel (888) 679-MERS M&I Bank FSB :
("Lender") 1s organized and existing under the laws of the United States o

All three models produced different chunks overall with BGE and E5 having more contextual aware answers. We had only 2 documents in our RAG pipeline here, one for the loan worksheet and other for the title. Mini LLM worked well however it was not able to parse complex figures properly, for example when asked about monthly payments it gave answer as 1,869.37, which was only one part of overall monthly payments as we can see below. Others answered 2,308.95 properly.

However, there was another query where we had conficting answers between Mini LLM and other 2 models, query asked for total amount of loan and other 2 models extracted answer from title deed while Mini LLM extracted answer from loan worksheet, both answers were different most likely because of error in documents or some other reason. Most likely these cases will get resolved as we add more documents in our RAG pipeline, overall we are indeed seeing better results when we are using more complex embedding models.

In [ ]:
retriever = index.as_retriever(similarity_top_k=10)
retrieved_nodes = retriever.retrieve(query)

In [ ]:
query = "What is the loan amount?"

# Try different values of top_k
for top_k in [2, 5, 10]:
    print(f"\n--- Results for top_k = {top_k} ---\n")
    retriever = index.as_retriever(similarity_top_k=top_k)
    nodes = retriever.retrieve(query)
    for i, node in enumerate(nodes):
        print(f"Result {i+1}:")
        print(node.get_text())
        print("-" * 80)



--- Results for top_k = 2 ---

Result 1:
Your actual rate, payment, and cost could be higher. Get an official Loan Estimate before choosing a loan.
Fee Details and Summary
Applicants: Application No:
Date Prepared:
Loan Program:
Prepared By:
THIS IS NOT A GOOD FAITH ESTIMATE (GFE). This "Fees W orksheet" is provided for informational purposes ONLY, to assist
you in determining an estimate of cash that may be required to close and an estimate of your proposed monthly mortgage 
payment. Actual charges may be more or less, and your transaction may not involve a fee for every item listed.
Total Loan Amount:  Interest Rate: Term/Due In:
Fee Paid To Paid By (Fee Split**) Amount PFC / F / POC
TOTAL ESTIMATED FUNDS NEEDED TO CLOSE: TOTAL ESTIMATED MONTHLY PAYMENT:
Total Estimated Funds Total Monthly Payment
Purchase Price (+)
Alterations (+)
Land (+)
Refi (incl. debts to be paid off) (+)
Est. Prepaid Items/Reserves (+)
Est. Closing Costs (+)
Loan Amount (-) Principal & Interest
Other Financin

In [ ]:
# Set a threshold (you can try 0.7, 0.75, 0.8, etc.)
threshold = 0.82

# Filter nodes based on score
filtered_nodes = [node for node in retrieved_nodes if node.score and node.score > threshold]

print(f"\nFiltered {len(filtered_nodes)} out of {len(retrieved_nodes)} total nodes.")

for i, node in enumerate(filtered_nodes):
    print(f"\nResult {i+1}: (Score: {node.score:.2f})")
    print(node.get_text())
    print("-" * 80)



Filtered 1 out of 4 total nodes.

Result 1: (Score: 0.84)
Your actual rate, payment, and cost could be higher. Get an official Loan Estimate before choosing a loan.
Fee Details and Summary
Applicants: Application No:
Date Prepared:
Loan Program:
Prepared By:
THIS IS NOT A GOOD FAITH ESTIMATE (GFE). This "Fees W orksheet" is provided for informational purposes ONLY, to assist
you in determining an estimate of cash that may be required to close and an estimate of your proposed monthly mortgage 
payment. Actual charges may be more or less, and your transaction may not involve a fee for every item listed.
Total Loan Amount:  Interest Rate: Term/Due In:
Fee Paid To Paid By (Fee Split**) Amount PFC / F / POC
TOTAL ESTIMATED FUNDS NEEDED TO CLOSE: TOTAL ESTIMATED MONTHLY PAYMENT:
Total Estimated Funds Total Monthly Payment
Purchase Price (+)
Alterations (+)
Land (+)
Refi (incl. debts to be paid off) (+)
Est. Prepaid Items/Reserves (+)
Est. Closing Costs (+)
Loan Amount (-) Principal & Intere

In [ ]:
experiments = [
    {"top_k": 5, "threshold": None},
    {"top_k": 8, "threshold": 0.75},
    {"top_k": 3, "threshold": 0.8},
]

In [ ]:
for exp in experiments:
    print(f"\\n--- Experiment: top_k={exp['top_k']}, threshold={exp['threshold']} ---")
    retriever = index.as_retriever(similarity_top_k=exp["top_k"])
    nodes = retriever.retrieve("What is the name of the borrower and the property address?")
    if exp["threshold"]:
        nodes = [node for node in nodes if node.score and node.score > exp["threshold"]]
    print(f"Chunks Retrieved: {len(nodes)}")
    for i, node in enumerate(nodes):
        print(f"Chunk {i+1} (Score: {node.score:.2f}):")
        print(node.get_text())
        print("-" * 80)

\n--- Experiment: top_k=5, threshold=None ---
Chunks Retrieved: 4
Chunk 1 (Score: 0.83):
which has the address of 6468 SOUTH 20TH STREET [Street]
MILWAUKEE [City], Wisconsin 53221 [Zip Code] ("Property Address"),

TOGETHER WITH all the improvements now or hereafter erected on the property, and all easements,
appurtenances and fixtures now or hereafter a part of the property All replacements and additions shall also be
covered by this Security Instrument All of the foregoing 1s referred to in this Security Instrument as the "Property "
Borrower understands and agrees that MERS holds only legal title to the mterests granted by Borrower in this
Security Instrument, but, 1f necessary to comply with law or custom, MERS, (as nominee for Lender and Lender's
successors and assigns), has the right to exercise any or all of those interests, including, but not limited to, the right
to foreclose and sell the Property, and to take any action required of Lender including, but not limited to, releasi

While we used custom retrievers in our rag pipeline we can observe the effect of Top K and threshold on the chunk selections even in default retriever. However in our case when only few chunks were available we are not seeing much change overall. And provided that we used reranking in our custom RAG pipeline where more score is given to more relevant node.